Source/towners.cpp

```diff
@@ -683,624 +680,693 @@ void ProcessTowners()
-					quests[Q_ROCK]._qactive = QUEST_DONE;
-					quests[Q_ROCK]._qvar2 = 2;
-					quests[Q_ROCK]._qvar1 = 2;
-					RemoveInvItem(p, i);
-					SpawnUnique(UITEM_INFRARING, towner[t]._tx, towner[t]._ty + 1);
-					towner[t]._tbtcnt = true;
-					towner[t]._tTalkingToPlayer = p;
-					InitQTextMsg(TEXT_INFRA7);
-					towner[t]._tMsgSaid = true;
+	}
+	if (!gbIsMultiplayer) {
+		if (player->_pLvlVisited[3] && quests[Q_LTBANNER]._qactive != QUEST_NOTAVAIL) {
+			if ((quests[Q_LTBANNER]._qactive == QUEST_INIT || quests[Q_LTBANNER]._qactive == QUEST_ACTIVE) && quests[Q_LTBANNER]._qvar2 == 0 && !towner->_tMsgSaid) {
+				quests[Q_LTBANNER]._qvar2 = 1;
+				if (quests[Q_LTBANNER]._qactive == QUEST_INIT) {
+					quests[Q_LTBANNER]._qvar1 = 1;
+					quests[Q_LTBANNER]._qactive = QUEST_ACTIVE;
 				}
+				quests[Q_LTBANNER]._qlog = true;
+				towner->_tbtcnt = true;
+				towner->_tTalkingToPlayer = player;
+				InitQTextMsg(TEXT_BANNER2);
+				towner->_tMsgSaid = true;
 			}
-			if (plr[p]._pLvlVisited[9] && quests[Q_ANVIL]._qactive != QUEST_NOTAVAIL) {
-				if ((quests[Q_ANVIL]._qactive == QUEST_INIT || quests[Q_ANVIL]._qactive == QUEST_ACTIVE) && quests[Q_ANVIL]._qvar2 == 0 && !towner[t]._tMsgSaid) {
-					if (quests[Q_ROCK]._qvar2 == 2 || (quests[Q_ROCK]._qactive == QUEST_ACTIVE && quests[Q_ROCK]._qvar2 == 1)) {
-						quests[Q_ANVIL]._qvar2 = 1;
-						quests[Q_ANVIL]._qlog = true;
-						if (quests[Q_ANVIL]._qactive == QUEST_INIT) {
-							quests[Q_ANVIL]._qactive = QUEST_ACTIVE;
-							quests[Q_ANVIL]._qvar1 = 1;
-						}
-						towner[t]._tbtcnt = true;
-						towner[t]._tTalkingToPlayer = p;
-						InitQTextMsg(TEXT_ANVIL5);
-						towner[t]._tMsgSaid = true;
-					}
-				}
-				if (quests[Q_ANVIL]._qvar2 == 1 && PlrHasItem(p, IDI_ANVIL, &i) != NULL) {
-					if (!towner[t]._tMsgSaid) {
-						quests[Q_ANVIL]._qactive = QUEST_DONE;
-						quests[Q_ANVIL]._qvar2 = 2;
-						quests[Q_ANVIL]._qvar1 = 2;
-						RemoveInvItem(p, i);
-						SpawnUnique(UITEM_GRISWOLD, towner[t]._tx, towner[t]._ty + 1);
-						towner[t]._tbtcnt = true;
-						towner[t]._tTalkingToPlayer = p;
-						InitQTextMsg(TEXT_ANVIL7);
-						towner[t]._tMsgSaid = true;
-					}
-				}
+			int i = player->GetItemIndex(IDI_BANNER);
+			if (quests[Q_LTBANNER]._qvar2 == 1 && i != -1 && !towner->_tMsgSaid) {
+				quests[Q_LTBANNER]._qactive = QUEST_DONE;
+				quests[Q_LTBANNER]._qvar1 = 3;
+				RemoveInvItem(player, i);
+				SpawnUnique(UITEM_HARCREST, towner->_tx, towner->_ty + 1);
+				towner->_tbtcnt = true;
+				towner->_tTalkingToPlayer = player;
+				InitQTextMsg(TEXT_BANNER3);
+				towner->_tMsgSaid = true;
 			}
 		}
-		if (!qtextflag) {
-			TownerTalk(TEXT_GRISWOLD1, t);
-			if (storeflag) {
-				StartStore(STORE_SMITH);
-			}
+	}
+	if (qtextflag)
+		return;
+	TownerTalk(TEXT_OGDEN1);
+	if (storeflag) {
+		StartStore(STORE_TAVERN);
+	}
+}
+
+void TalkToDeadguy(PlayerStruct *player, TownerStruct *towner)

```

Comment:
Seems weird to me to have the towner name in the function name, yet take a `TownerStruct` as parameter. What if someone calls this method with a towner other than the deadguy? 

I think this is a design smell.

Reply:
It's a type, there could potentially be multiple dead guy towners. We could mabe handle this as a towner-ai, to mirror monsters even more.

Reply:
> there could potentially be multiple dead guy towners

This seems a bit farfetched to me considering what we currently have. Right now, looking from the outside, it would make more sense for these methods to _not_ take a `TownerStruct` argument and instead each method pick the correct towner from the global towner array.

Obviously that's _also_ less than ideal... this looks like a proper case of polimorphism to me where each towner implementation would do something different on the same method (or something like that).

Reply:
there are 3 towners wo all have the cow type, also Wirt's id doesnt align with his type.

I dontsee an issue with my suggestion to turn this in to towner AIs?